pages/instructorFileBrowser/instructorFileBrowser.js

```diff
@@ -202,89 +181,74 @@ function browseDirectory(file_browser, callback) {
-                  encodedName: encodePath(filename),
-                  path: path.relative(file_browser.paths.coursePath, filepath),
-                  encodedPath: encodePath(path.relative(file_browser.paths.coursePath, filepath)),
-                  canView: !file_browser.paths.invalidRootPaths.some((invalidRootPath) =>
-                    contains(invalidRootPath, filepath)
-                  ),
-                });
-              }
-              callback(null);
-            });
-          },
-          (err) => {
-            if (ERR(err, callback)) return;
-            callback(null);
+            const stats = await fs.lstat(filepath);
+            if (stats.isFile()) {
+              const editable = !(await isBinaryFile(filepath));
+              const movable = !file_browser.paths.cannotMove.includes(filepath);
+              return {
+                id: index,
+                name: filename,
+                isFile: true,
+                encodedName: encodePath(filename),
+                path: path.relative(file_browser.paths.coursePath, filepath),
+                encodedPath: encodePath(path.relative(file_browser.paths.coursePath, filepath)),
+                dir: file_browser.paths.workingPath,
+                canEdit:
+                  editable &&
+                  file_browser.has_course_permission_edit &&
+                  !file_browser.example_course,
+                canUpload: file_browser.has_course_permission_edit && !file_browser.example_course,
+                canDownload: true, // we already know the user is a course Viewer (checked on GET)
+                canRename:
+                  movable &&
+                  file_browser.has_course_permission_edit &&
+                  !file_browser.example_course,
+                canDelete:
+                  movable &&
+                  file_browser.has_course_permission_edit &&
+                  !file_browser.example_course,
+                canView: !file_browser.paths.invalidRootPaths.some((invalidRootPath) =>
+                  contains(invalidRootPath, filepath)
+                ),
+              };
+            } else if (stats.isDirectory()) {
+              return {
+                id: index,
+                name: filename,
+                isDirectory: true,
+                encodedName: encodePath(filename),
+                path: path.relative(file_browser.paths.coursePath, filepath),
+                encodedPath: encodePath(path.relative(file_browser.paths.coursePath, filepath)),
+                canView: !file_browser.paths.invalidRootPaths.some((invalidRootPath) =>
+                  contains(invalidRootPath, filepath)
+                ),
+              };
+            } else {
+              return null;
+            }
           }
         );
+        file_browser.files = all_files.filter((f) => f?.isFile).sort((f) => f.name);
+        file_browser.dirs = all_files.filter((f) => f?.isDirectory).sort((f) => f.name);
```

Comment:
`Array.prototype.sort` takes a comparison function, not something that maps to a property to be sorted on. I think this is what you want (untested):

```suggestion
        file_browser.files = all_files.filter((f) => f?.isFile).sort((a, b) => a.name.localeCompare(b.name));
        file_browser.dirs = all_files.filter((f) => f?.isDirectory).sort((a, b) => a.name.localeCompare(b.name));
```

Reply:
Indeed, I was thinking of the Python [sort](https://docs.python.org/3/library/stdtypes.html#list.sort) function.